lib/checkother.cpp

```diff
@@ -1391,6 +1391,59 @@ void CheckOther::checkConstVariable()
         // Skip if address is taken
         if (Token::findmatch(var->nameToken(), "& %varid%", scope->bodyEnd, var->declarationId()))
             continue;
+        // Skip if another non-const variable is initialized with this variable
+        {
+            //Is it the right side of an initialization of a non-const reference
+            bool usedInAssignment = [&]
+                {
+                    auto findAssignment = [&](const Token* t) { return Token::findmatch(t, "%type% & %name% = %varid%", scope->bodyEnd, var->declarationId()); };
+                    for (const Token* match = findAssignment(var->nameToken()); match != nullptr; match = findAssignment(match->next()))
+                    {
+                        const Variable* matchvar = match ? match->tokAt(2)->variable() : nullptr;
+                        if (matchvar && !matchvar->isConst())
+                            return true;
+                    }
+                    return false;
+                }();
+            if (usedInAssignment)
+                continue;
+        }
+        // Skip if we ever cast this variable to a non-const type
+        {
+            bool castToNonConst = [&]
+                {
+                    for (const Token* tok = var->nameToken(); tok != scope->bodyEnd && tok != nullptr; tok = tok->next())
+                    {
+                        if (tok->isCast())
+                        {
+                            if (Token::simpleMatch(tok->previous(), ">"))
+                            {
+                                // dynamic_cast<X&>(y)
+                                //                 ^    we're at this operator
+                                bool castToConst = Token::simpleMatch(tok->previous()->link()->next(), "const");
+                                if (!castToConst)
+                                {
+                                    return true;
+                                }
+                            }
+                            else
+                            {
+                                // (X&)(y)
+                                // ^        we're at this operator
+                                bool castToConst = Token::simpleMatch(tok->next(), "const");
```

Comment:
I believe, if `valueType` is used we shouldn't need separate handling for C and C++ casts.